src/shogun/optimization/StanFirstOrderSAGCostFunction.h

```diff
@@ -0,0 +1,177 @@
+          REQUIRE(y.size() > 0, "Empty y provided");
+          auto num_of_variables = trainable_parameters.rows();
+          REQUIRE(
+              num_of_variables > 0, "Provided %d variables in the parameters, more "
+                                    "than 0 parameters required",
+              num_of_variables);
+
+          m_X = X;
+          m_y = y;
+          m_cost_for_ith_point = cost_for_ith_point;
+          m_total_cost = total_cost;
+          m_ref_trainable_parameters = SGVector<float64_t>(num_of_variables);
+          for (auto i : range(num_of_variables))
+          {
+            m_ref_trainable_parameters[i] = m_trainable_parameters(i, 0).val();
+          }
+        }
+
+		/** Setter for the training data X */
+		virtual void
+		set_training_data(SGMatrix<float64_t> X_new, SGMatrix<float64_t> y_new);
+
+		virtual ~StanFirstOrderSAGCostFunction();
+
+		/** Initialize to generate a sample sequence
+		 *
+		 */
+		virtual void begin_sample();
+
+		/** Get next sample
+		 *
+		 * @return false if reach the end of the sample sequence
+		 * */
+		virtual bool next_sample();
+
+		/** Get the sample gradient value wrt target variables
+		 *
+		 * WARNING
+		 * This method does return
+		 * \f$ \frac{\partial f_i(w) }{\partial w} \f$,
+		 * instead of
+		 * \f$\sum_i{ \frac{\partial f_i(w) }{\partial w} }\f$
+		 *
+		 * For least squares cost function, that is the value of
+		 * \f$\frac{\partial f_i(w) }{\partial w}\f$ given \f$w\f$ is known
+		 * where the index \f$i\f$ is obtained by next_sample()
+		 *
+		 * @return sample gradient of variables
+		 */
+		virtual SGVector<float64_t> get_gradient();
+
+		/** Get the cost given current target variables
+		 *
+		 * For least squares, that is the value of \f$f(w)\f$.
+		 *
+		 * @return cost
+		 */
+		virtual float64_t get_cost();
+
+		/** Get the sample size
+		 *
+		 * @return the sample size
+		 */
+		virtual index_t get_sample_size();
+
+		/** Get the average gradient value wrt target variables
+		 *
+		 * Note that the average gradient is the mean of sample gradient from
+		 * get_gradient()
+		 * if samples are generated (uniformly) at random.
+		 *
+		 * WARNING
+		 * This method returns
+		 * \f$ \frac{\sum_i^n{ \frac{\partial f_i(w) }{\partial w} }}{n}\f$
+		 *
+		 * For least squares, that is the value of
+		 * \f$ \frac{\frac{\partial f(w) }{\partial w}}{n} \f$ given \f$w\f$ is
+		 * known
+		 * where \f$f(w)=\frac{ \sum_i^n{ (y_i-w^t x_i)^2 } }{2}\f$
+		 *
+		 * @return average gradient of target variables
+		 */
+		virtual SGVector<float64_t> get_average_gradient();
+
+		virtual SGVector<float64_t> obtain_variable_reference();
+
+		/** Updates m_trainable_parameters values to m_ref_trainable_parameters
+		 */
+		void update_stan_vectors_to_reference_values();
+
+	protected:
+		/** X is the training data in column major matrix format */
+		SGMatrix<float64_t> m_X;
+
+		/** y is the ground truth, or the correct prediction */
+		SGMatrix<float64_t> m_y;
```

Comment:
`FirstOrderSAGCostFunction` does not have members for the training data. Why are they needed here?

Reply:
@iglesias In the unit test of FirstOrderSAGCostFunction, we implement an additional class called CRegressionExample that wraps FirstOrderSAGCostFunction and contains the training data. I've simply removed the necessity for that class since it just acted as a wrapper around the cost function and data, and have included the data in the loss function, or atleast a reference to the data.

Reply:
We must be careful and we must no add unnecessary relationships. If it turns out that the training data is a member both in the cost function, and in the neural network, and in etc etc, that is going to cause lot of usage confusion.

It does not sounds unreasonable that the tests have a facility packaged to prepare input data and avoid code duplication (e.g. inside a class a the CRegressionExample you are mentioning). What did you find wrong with it?